OpcUaClient/OpcUaInterface.hpp

```diff
@@ -218,14 +232,193 @@ class OpcUaWrapper : public OpcUaInterface {
+					if(pWrapper != nullptr && pWrapper->eventcallback != nullptr) {
+						for(int i = 0; i < eventField.arrayLength; i++) {
+							UA_ModelChangeStructureDataType modelChangeStructureDataType = modelChangeStructureDataTypes[i];
+							UA_NodeId affectedNode = modelChangeStructureDataType.affected;
+							UA_NodeId affectedType = modelChangeStructureDataType.affectedType;
+							UA_Byte verb = modelChangeStructureDataType.verb;
+							Umati::Dashboard::IDashboardDataClient::StructureChangeEvent stc;
+							ModelOpcUa::NodeId_t nodeId = ModelOpcUa::NodeId_t();
+							nodeId.Uri = pWrapper->namespaceArray.at(affectedNode.namespaceIndex);
+							std::string nodeIdPrefix = "i=";
+							switch(affectedNode.identifierType){
+								case UA_NODEIDTYPE_NUMERIC: nodeIdPrefix = "i="; break;
+								case UA_NODEIDTYPE_STRING: nodeIdPrefix = "s="; break;
+								case UA_NODEIDTYPE_BYTESTRING: nodeIdPrefix = "b="; break;
+								case UA_NODEIDTYPE_GUID: nodeIdPrefix = "g="; break;
+							}
+							nodeId.Id = nodeIdPrefix + std::to_string(affectedNode.identifier.numeric);
+							stc.refreshNode = nodeId;
+							stc.nodeAdded = (verb & UA_MODELCHANGESTRUCTUREVERBMASK_NODEADDED) == UA_MODELCHANGESTRUCTUREVERBMASK_NODEADDED;
+							stc.nodeDeleted = (verb & UA_MODELCHANGESTRUCTUREVERBMASK_NODEDELETED) == UA_MODELCHANGESTRUCTUREVERBMASK_NODEDELETED;
+							stc.referenceAdded = (verb & UA_MODELCHANGESTRUCTUREVERBMASK_REFERENCEADDED) == UA_MODELCHANGESTRUCTUREVERBMASK_REFERENCEADDED;
+							stc.referenceDeleted = (verb & UA_MODELCHANGESTRUCTUREVERBMASK_REFERENCEDELETED) == UA_MODELCHANGESTRUCTUREVERBMASK_REFERENCEDELETED;
+							stc.dataTypeChanged = (verb & UA_MODELCHANGESTRUCTUREVERBMASK_DATATYPECHANGED) == UA_MODELCHANGESTRUCTUREVERBMASK_DATATYPECHANGED;
+							pWrapper->eventcallback(stc);
+						}	
+					} else {
+						LOG(ERROR) << "Unable to propagate Event callback!";
+					}
+					UA_Array_delete(modelChangeStructureDataTypes, eventField.arrayLength, &UA_TYPES[UA_TYPES_MODELCHANGESTRUCTUREDATATYPE]);
+				}
+			}
+		}
+			std::shared_ptr<Dashboard::IDashboardDataClient::EventSubscriptionHandle> EventSubscribe(UA_Client* client,
+			Dashboard::IDashboardDataClient::eventCallbackFunction_t eventcallback) {
+				this->eventcallback = eventcallback;
+				UA_CreateSubscriptionRequest request = UA_CreateSubscriptionRequest_default();
+    			UA_CreateSubscriptionResponse response = UA_Client_Subscriptions_create(client, request, NULL, NULL, NULL);
+				if(response.responseHeader.serviceResult == UA_STATUSCODE_GOOD) {
+					UA_UInt32 subId = response.subscriptionId;
+					UA_MonitoredItemCreateRequest item;
+    				UA_MonitoredItemCreateRequest_init(&item);
+					item.itemToMonitor.attributeId = UA_ATTRIBUTEID_EVENTNOTIFIER;
+					item.monitoringMode = UA_MONITORINGMODE_REPORTING;
+    				item.itemToMonitor.nodeId = UA_NODEID_NUMERIC(0, 2253); // Root->Objects->Server
```

Comment:
can we use UA_NS0ID_SERVER instead of a the number?

Reply:
Done.